articles/external_qos_file.md

```diff
@@ -0,0 +1,402 @@
+
+In that way, the user loading the node is responsible of passing the desired QoS, instead of the node container.
+The message to be sent could be loaded from a qos profile file.
+
+#### Proposed interfaces
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
+```
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
+```
+
+```
+# composition_interfaces/srv/LoadNode.srv
+
+# Current request fields
+...
+# New field
+EndpointQoS[] requested_qos_profiles  # new field to allow loading user requested qos profiles
+---
+# Current response definition
+...
+```
+
+A `rcl_qos_loaded_profiles_t` can be constructed from the received message and passed to the node options of the node to be loaded.
+
+TBD: `QoS` message definition.
+
+### Support in launch files
+
+`launch_ros.actions.Node` and `launch_ros.actions.LoadComposableNode` actions could support a qos file argument in their constructors.
+
+## File format
+
+As showed in the above examples, both YAML or XML can correctly represent the desired file format.
+
+XML advantages:
+- Can be easily verified.
+YAML advantages:
+- It can be argued that the format is simpler and more readable than XML.
+- The format proposed in the above examples can easily be integrated with the current parameter file format.
+  That integration would lead in a unique "configuration file", that could be further extended in the future.
+  For example, passing remapping rules in it.
+
+## Explicitly allowing external configurability in entities
+
+Both node and init options could have a flag to ignore the new `--qos-file` argument.
+In a similar way, there could be an option for publishers/subscriptions/clients and services to allow side loading their qos or not.
+
+In nodes designed to be reused, it does make sense to allow side-loading the qos settings of all entities.
+To make this use case easier, the following options can be added:
+
+- `rcl_init_options_t` is extended with a `bool allow_side_loading_qos`.
+- The following enum is defined:
+  ```c
+  typedef enum rcl_side_load_qos_t {
+      RCL_SIDE_LOAD_QOS_DEFAULT;
+      RCL_SIDE_LOAD_QOS_ENABLED;
+      RCL_SIDE_LOAD_QOS_DISABLED;
+  } rcl_side_load_qos_t;
+  ```
+- A `rcl_side_load_qos_t allow_side_loading_qos` member is added to `rcl_node_options_t`, it will copy the behavior defined in the `rcl_init_options_t` of its parent context when the enum value is `RCL_SIDE_LOAD_QOS_DEFAULT`.
+- In a similar way, publisher/subscription/client/service options will have a `rcl_side_load_qos_t allow_side_loading_qos` that will copy the behavior set in the node in case `RCL_SIDE_LOAD_QOS_DEFAULT` is used.
```

Comment:
I think for individual entities, we will want to be able to turn on/off the configurability of specific QoS settings. For example, letting users change the history depth, but not the durability of a subscription.

I imagine an easy way to enable this capability is to allow node authors to register a callback (like you mention at the end of this document). The callback gets called when users try to configure a QoS setting (or group of QoS settings); similar to "on set" parameter callbacks, they can return "SUCCESS" or "REJECT" based on whatever logic they like. I guess this has the downside of not being easily implemented at the `rcl` layer. Perhaps a more static description of what QoS settings are allowed to be configured is better for `rcl`. QoS settings have a set well-known options after all.

Reply:
> I think for individual entities, we will want to be able to turn on/off the configurability of specific QoS settings. For example, letting users change the history depth, but not the durability of a subscription.

I don't see when not allowing to change the durability policy is useful.

Reply:
> I don't see when not allowing to change the durability policy is useful.

We can imagine a scenario where the node author does *not* want a subscription to get messages that were published before it started (e.g. don't allow *transient_local*). If an integrator is allowed to change the durability, they may be violating the node authors assumptions, and inadvertently "break" the node. I think having this fine-grained ability to turn on/off different QoS settings makes it easier for the node author to set expectations about how the node is meant to be used.

One example that comes to mind is the [static TF broadcaster](https://github.com/ros2/geometry2/blob/e813f131e5663133b5db956fae1f413b6290a303/tf2_ros/include/tf2_ros/static_transform_broadcaster_node.hpp) and TF listener, which uses *transient local* for it's durability on the `/tf_static` topic. If we let an integrator change the durability to *volatile*, it would break assumptions for a lot of tooling built around the `/tf_static` topic.

Reply:
I've added the concept QoS acceptance callbacks to the proposal https://github.com/ros2/design/pull/296/commits/a0e635f3500e457bf56a36862651fe3bcf50b0a7.